src/components/tl/ucp/gather/gather_knomial.c

```diff
@@ -9,135 +9,195 @@
+                } else {
+                    /*
+                     the root is a particular case because it must aggregate
+                     the data sorted by ranks
+                    */
+                    scratch_offset = PTR_OFFSET(task->gather_kn.scratch,
+                                                data_size * peer);
+                    num_blocks = ucc_min(task->gather_kn.dist, tsize - vpeer);
+                    /* check if received data correspond to contiguous ranks */
+                    if ((ct == UCC_COLL_TYPE_REDUCE) ||
+                        (num_blocks <= tsize - peer)) {
+                        msg_size = peer_seg_count * dt_size;
+                        if (args->coll_type != UCC_COLL_TYPE_GATHER) {
+                            scratch_offset = PTR_OFFSET(task->gather_kn.scratch,
+                                                        peer_seg_offset * dt_size);
                         }
+                        UCPCHECK_GOTO(ucc_tl_ucp_recv_nb(scratch_offset,
+                                                         msg_size, mtype,
+                                                         peer, team, task),
+                                        task, out);
+                    } else {
+                        /*
+                         in this case, data must be split in two
+                         at the destination buffer
+                        */
+                        msg_size = data_size * (tsize - peer);
+                        UCPCHECK_GOTO(ucc_tl_ucp_recv_nb(scratch_offset,
+                                                         msg_size, mtype,
+                                                         peer, team, task),
+                                      task, out);
+                        msg_size = data_size * (num_blocks - (tsize - peer));
+                        UCPCHECK_GOTO(ucc_tl_ucp_recv_nb(task->gather_kn.scratch,
+                                                         msg_size, mtype,
+                                                         peer, team, task),
+                                      task, out);
                     }
                 }
+            }
 
-                if (task->gather_kn.dist == 1) { //check if first passage
-                    msg_size = data_size;
-                    if (rank != root) {
-                        status = ucc_mc_memcpy(task->gather_kn.scratch,
-                                            args->src.info.buffer, msg_size,
-                                            args->src.info.mem_type, mtype);
-                    } else if (!UCC_IS_INPLACE(*args)) {
-                        status = ucc_mc_memcpy(
-                            PTR_OFFSET(task->gather_kn.scratch, data_size * rank),
-                            args->src.info.buffer, msg_size,
-                            args->src.info.mem_type, mtype);
-                    }
+            if ((ct != UCC_COLL_TYPE_REDUCE)  &&
```

Comment:
Please add a comment why reduce collective is checked for in a gather related file.